.travis.yml

```diff
@@ -7,30 +7,39 @@ notifications:
 
 # Environment setup before test script. Runs for each build
 before_install:
-  # Check if anything has changed within the docker directory
-  - DOCKER_CHANGED=`git diff --name-only $TRAVIS_COMMIT_RANGE -- tools/docker | wc -l`
-  # If Docker directory has not changed, pull image from Dockerhub. Else, build
-  # image from Dockerifle. This needs to be done for each job. Any build error
-  # will count as Travis test failure. In case this updates develop, push new
-  # image to Dockerhub (secure credentials only readable on bulids to
-  # contiki-ng/contiki-ng branches, not forks or PRs)
+  # Construct the correct container image tag corresponding to this branch/pull
+  - CONTIKER_TAG=`git log -1 --oneline -- tools/docker/ | cut -d" " -f1`
+  - DOCKER_IMG=$DOCKER_BASE_IMG:$CONTIKER_TAG
+  # Try to download the image from dockerhub. If it works, use it.
+  #
+  # If however it fails then we are most likely looking at a branch or pull
+  # that touched tools/docker. In this case, build the image.
+  #
+  # Any build error will count as Travis test failure.
+  #
+  # If the test was triggered by a branch update (e.g. a PR merge) then push
+  # the new image to dockerhub. This will only happen for builds against
+  # contiki-ng/contiki-ng, not for builds on forks.
+  - echo $DOCKER_IMG
   - >
-    if [ $DOCKER_CHANGED == 0 ]; then
-      echo "Docker image unchanged, pull from Dockerhub"
-      docker pull $DOCKER_IMG;
-    else
-      echo "Docker image changed, build from Dockerfile"
+    echo "Pulling image $DOCKER_IMG from dockerhub";
+    docker pull $DOCKER_IMG;
+    if [ $? != 0 ]; then
+      echo $CONTIKER_TAG "does not exist or pull failed";
+      echo "This is normal for PR builds and for builds on forks";
+      echo "Building from dockerfile";
       docker build tools/docker -t $DOCKER_IMG;
       if [ $? != 0 ]; then
         echo "Failed to build Docker image"
         exit 1
       fi
-      if [ $TRAVIS_SECURE_ENV_VARS == true ] && [ $TRAVIS_PULL_REQUEST == false ] && [ $TRAVIS_BRANCH == 'develop' ]; then
-        echo "This build is for an update of branch develop. Push image to Dockerhub"
+      if [ $TRAVIS_REPO_SLUG == contiki-ng/contiki-ng ] && [ $TRAVIS_PULL_REQUEST == false ]; then
```

Comment:
why the change from `TRAVIS_SECURE_ENV_VARS`?

Reply:
As per the OP: `$TRAVIS_SECURE_ENV_VARS` can be true outside `contiki-ng/contiki-ng` even though the actual credentials are not available

Reply:
Sorry I was lazy and read the code not the text :p

Reply:
Maybe what we want is check both vars?

Reply:
I hope the secure creds are indeed only available where we want them to...

Reply:
So this is what I saw on a [branch build](https://travis-ci.org/g-oikonomou/contiki-ng/jobs/603332312) under `g-oikonomou/contiki-ng`

```
$ echo "Pulling image $DOCKER_IMG from dockerhub"; docker pull $DOCKER_IMG; if [ $? != 0 ]; then
    echo $CONTIKER_TAG " does not exist or pull failed.";
    echo "Building from dockerfile";
    docker build tools/docker -t $DOCKER_IMG;
    if [ $? != 0 ]; then
      echo "Failed to build Docker image"
      exit 1
    fi
    echo "Would push $DOCKER_IMG";
    if [ $TRAVIS_SECURE_ENV_VARS == true ] && [ $TRAVIS_PULL_REQUEST == false ]; then
      echo "This is a branch build that updates the container. Push image to Dockerhub"
      echo $DOCKERHUB_PASSWD | docker login --username contiker --password-stdin
      docker push $DOCKER_IMG;
    fi
  fi

Pulling image contiker/contiki-ng:afdedad from dockerhub
Error response from daemon: manifest for contiker/contiki-ng:afdedad not found
afdedad  does not exist or pull failed.
Building from dockerfile
Sending build context to Docker daemon  6.144kB


Step 1/23 : FROM 32bit/ubuntu:16.04
[...]
Step 23/23 : CMD bash --login
 ---> Running in c3c9614ed927
 ---> d93b180c1c99
Removing intermediate container c3c9614ed927
Successfully built d93b180c1c99
Successfully tagged contiker/contiki-ng:afdedad
Would push contiker/contiki-ng:afdedad
This is a branch build that updates the container. Push image to Dockerhub
Error: Cannot perform an interactive login from a non TTY device
The push refers to a repository [docker.io/contiker/contiki-ng]
84d210f3adb0: Preparing
307f5dc1c117: Preparing
0074e8cb5ff8: Preparing
eb10466b0f03: Preparing
e91e41d09bfd: Preparing
9126e1fda065: Preparing
7d0b734d3160: Preparing
b008139cb6b0: Preparing
0dc69be94b41: Preparing
856263a8407c: Preparing
45ddea7619dd: Preparing
339fd2f0e879: Preparing
9126e1fda065: Waiting
7d0b734d3160: Waiting
b008139cb6b0: Waiting
0dc69be94b41: Waiting
856263a8407c: Waiting
45ddea7619dd: Waiting
339fd2f0e879: Waiting
denied: requested access to the resource is denied

Reply:
Right, looks all good. Looking at the docs on `TRAVIS_SECURE_ENV_VARS` it appears that this is set whenever secure vars are configured (regardless of whether available on a given job).